### PR TITLE
[Verification 9] Feature verification verifier comparing rendering results

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/243.
Resolves https://github.com/golemfactory/concent/issues/527.

### CODE REVIEW

File: concent_api/verifier/tests/test_integration_verification.py [hunk 1]
```diff
@@ -240,5 +272,200 @@ def test_blender_verification_order_should_call_verification_result_with_result_
...
+        self.assertEqual(mock_get_files_list_from_archive.call_count, 1)
+        self.assertEqual(mock_delete_file.call_count, 2)
+        mock_verification_result.assert_called_once_with(
+            self.compute_task_def['subtask_id'],
+            VerificationResult.ERROR,
+            'error',
+            ErrorCode.VERIFIIER_LOADING_FILES_INTO_MEMORY_FAILED.name,
+        )
+
+    def test_blender_verification_order_should_continue_verification_if_opening_first_file_raise_os_error(self):
```
Comment (dybi) [inline concent_api/verifier/tests/test_integration_verification.py]: BTW: I am not sure if this is proper behaviour - if reading the file fails and it cannot be uploaded, why should verification continue?

File: concent_api/verifier/tasks.py [hunk 2]
```diff
@@ -144,27 +152,125 @@ def blender_verification_order(
...
+        logger.error(f'Loading result files into memory failed with: {exception}')
+        verification_result.delay(
+            subtask_id,
+            VerificationResult.ERROR,
+            str(exception),
+            ErrorCode.VERIFIIER_LOADING_FILES_INTO_MEMORY_FAILED.name
+        )
+        return
+
+    if upload_file_content is not None:
```
Comment (dybi) [inline concent_api/verifier/tasks.py]: I would prefer not to have `None` based checks... This whole `if` block should be moved below line `182`

File: concent_api/verifier/tasks.py [hunk 3]
```diff
@@ -40,7 +48,7 @@ def blender_verification_order(
     result_size: int,
     result_package_hash: str,
     output_format: str,
-    scene_file: str,  # pylint: disable=unused-argument
+    scene_file: str,
```
Comment (dybi) [inline concent_api/verifier/tasks.py]: this whole task has grown dramatically. There's a lot of logic and for me it is very hard to follow what's going on. I would really instist on having it refactored in a way that small functions are extracted and the task consits of calls to such small functions.

File: concent_api/verifier/tasks.py [hunk 4]
```diff
@@ -144,27 +152,125 @@ def blender_verification_order(
...
-    )
+    blender_output_file_name = generate_blender_output_file_name(scene_file)
+    upload_file_name = generate_upload_file_name(subtask_id, output_format)
+
+    upload_file_content = None
+
+    try:
+        with open(generate_verifier_storage_file_path(blender_output_file_name), 'r') as upload_file:
+            upload_file_content = upload_file.read()
+    except OSError as exception:
```
Comment (dybi) [inline concent_api/verifier/tasks.py]: why processing is not stopped on `OSError`, while it is on `MemoryError`?

File: concent_api/utils/constants.py [hunk 5]
```diff
@@ -57,7 +57,11 @@ class ErrorCode(enum.Enum):
     QUEUE_WRONG_STATE                                                   = 'queue.wrong_state'
     REQUEST_BODY_NOT_EMPTY                                              = 'request_body.not_empty'
     SUBTASK_DUPLICATE_REQUEST                                           = 'subtask.duplicate_request'
+    VERIFIIER_COMPUTING_SSIM_FAILED                                     = 'verifier.computing_ssim_failed'
     VERIFIIER_FILE_DOWNLOAD_FAILED                                      = 'verifier.file_download_failed'
+    VERIFIIER_LOADING_FILES_INTO_MEMORY_FAILED                          = 'verifier.loading_files_into_memory_failed'
```
Comment (dybi) [inline concent_api/utils/constants.py]: I am mot sure if we need to distinguish between these two:
`VERIFIIER_LOADING_FILES_INTO_MEMORY_FAILED` vs `VERIFIIER_LOADING_FILES_WITH_OPENCV_FAILED`

File: concent_api/verifier/tests/test_integration_verification.py (hunk 1 above)
Comment (cameel) [reply concent_api/verifier/tests/test_integration_verification.py]: I agree. The verification should crash if the file cannot be opened.

The only case where we want to ignore problems with a file is when we try to upload our rendering result to the storage server. It's only for auditing purposes so we can ignore it and continue verification. Any other file is critical for us and a problem should result either in a crash (if we're sure this could have happened only due to a bug) or a verification error reported to the client (if it's client's fault, e.g. invalid file type or an empty file).

BTW, I see a typo in the code above: `VERIFIIER` -> `VERIFIER`.

File: concent_api/verifier/tasks.py (hunk 4 above)
Comment (cameel) [reply concent_api/verifier/tasks.py]: Yeah. I wouldn't catch `OSError` here. Make sure that the file exists. If it does than anything else going wrong is probably due to container misconfiguration, lack of disk space, etc.

File: concent_api/utils/constants.py (hunk 5 above)
Comment (cameel) [reply concent_api/utils/constants.py]: You have double `I` here.

File: concent_api/concent_api/tests/test_check_verifier_min_ssim.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        errors = check_verifier_min_ssim()
+
+        self.assertEqual(len(errors), 1)
+        self.assertEqual(errors[0], create_error_32_verifier_min_ssim_is_set())
+
+    @override_settings(
+        CONCENT_FEATURES=[],
+        VERIFIER_MIN_SSIM=None,
+    )
+    def test_that_verifier_min_ssim_set_will_not_produce_error_when_verifier_is_none_in_available_concent_features(self):
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/concent_api/tests/test_check_verifier_min_ssim.py]: Actually it was true too, but your version sounds much better :)

File: concent_api/verifier/tasks.py [hunk 7]
```diff
@@ -144,27 +152,125 @@ def blender_verification_order(
...
+            VerificationResult.ERROR,
+            str(exception),
+            ErrorCode.VERIFIIER_COMPUTING_SSIM_FAILED.name
+        )
+        return
+
+    assert isinstance(ssim, float)
+
+    # Compare SSIM with VERIFIER_MIN_SSIM.
+    if ssim > settings.VERIFIER_MIN_SSIM:
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: It is ok, from issue description:
`If the computed SSIM value is above the threshold, add verification_result with result == MATCH to the queue.`

Btw. I have reversed the condition to match our styling.

File: concent_api/verifier/tests/test_integration_verification.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/verifier/tests/test_integration_verification.py]: It is working ok, just the test name was misleading, `first_file` meant blender output file in this context, I have changed the test name.

File: concent_api/verifier/tasks.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: @dybi Lets discuss this tomorrow.

File: concent_api/utils/constants.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/utils/constants.py]: Right, I will get rid of second one.

File: concent_api/verifier/tasks.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: @dybi I have followed @cameel comment in https://github.com/golemfactory/concent/pull/318#discussion_r195136617

```
This should be outside of the try ... except block. An exception handler should cover only the code where you expect the exception.
```

Personally I think both of you are right in this case, but using `None` is less evil. I have moved place where variable is created to `expect` block

File: concent_api/verifier/tests/test_integration_verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/verifier/tests/test_integration_verification.py]: :+1: done

File: concent_api/utils/constants.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/utils/constants.py]: :+1: done

File: concent_api/verifier/tasks.py [hunk 8]
```diff
@@ -119,12 +127,12 @@ def blender_verification_order(
             unpack_archive(
                 os.path.basename(file_path)
             )
-        except (OSError, BadZipFile) as e:
+        except (OSError, BadZipFile) as exception:
```
Comment (rwrzesien) [inline concent_api/verifier/tasks.py]: @cameel comment from https://github.com/golemfactory/concent/pull/514#discussion_r195779279
> What kind of OSError are you expecting here? I think you should crash on OSError.

Any problem that is related with reading file from disk. I have followed info from issue that if archive file can't be extracted (in this case, it can't too) verification should end here.

File: concent_api/verifier/tasks.py [hunk 9]
```diff
@@ -119,12 +127,12 @@ def blender_verification_order(
             unpack_archive(
```
Comment (rwrzesien) [inline concent_api/verifier/tasks.py]: @cameel comment from https://github.com/golemfactory/concent/pull/514#discussion_r195780046

> Will it crash if any of the unpacked files already exists? It should. We prefer to crash than to overwrite anything.

File: concent_api/verifier/tasks.py [hunk 10]
```diff
@@ -100,7 +108,7 @@ def blender_verification_order(
                 subtask_id,
                 VerificationResult.ERROR.name,
                 str(exception),
-                ErrorCode.VERIFIIER_FILE_DOWNLOAD_FAILED.name
```
Comment (rwrzesien) [inline concent_api/verifier/tasks.py]: @cameel comment from https://github.com/golemfactory/concent/pull/514#discussion_r195780493

> Will it crash if the file being downloaded already exists? It should.
> Any file operation where the file being created is not supposed to exist but it does should crash the verifier so that we can detect and fix the issue.

File: concent_api/concent_api/settings/base.py [hunk 11]
```diff
@@ -223,6 +223,12 @@
             'level':     'DEBUG',
             'propagate': True,
         },
+        'concent.crash': {
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/concent_api/settings/base.py]: Please check my answer here https://github.com/golemfactory/concent/pull/518#discussion_r196187130

File: concent_api/verifier/tasks.py [hunk 12]
```diff
@@ -144,27 +152,126 @@ def blender_verification_order(
...
-        VerificationResult.MATCH.name,
-    )
+    blender_output_file_name = generate_blender_output_file_name(scene_file)
+    upload_file_name = generate_upload_file_name(subtask_id, output_format)
+
+    # Read Blender output file.
+    try:
+        with open(generate_verifier_storage_file_path(blender_output_file_name), 'r') as upload_file:
+            upload_file_content = upload_file.read()
+    except OSError as exception:
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: Ok, however the issue description says that we should continue.

File: concent_api/verifier/tasks.py (hunk 3 above)
Comment (dybi) [reply concent_api/verifier/tasks.py]: as discussed it will be done in separate issue

Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: Issue https://github.com/golemfactory/concent/issues/524

File: concent_api/utils/constants.py [hunk 13]
```diff
@@ -58,8 +58,12 @@ class ErrorCode(enum.Enum):
...
     REQUEST_BODY_NOT_EMPTY                                              = 'request_body.not_empty'
     SUBTASK_DUPLICATE_REQUEST                                           = 'subtask.duplicate_request'
-    VERIFIIER_FILE_DOWNLOAD_FAILED                                      = 'verifier.file_download_failed'
-    VERIFIIER_UNPACKING_ARCHIVE_FAILED                                  = 'verifier.unpacking_archive_failed'
+    VERIFIER_COMPUTING_SSIM_FAILED                                     = 'verifier.computing_ssim_failed'
+    VERIFIER_FILE_DOWNLOAD_FAILED                                      = 'verifier.file_download_failed'
+    VERIFIER_LOADING_FILES_INTO_MEMORY_FAILED                          = 'verifier.loading_files_into_memory_failed'
+    VERIFIER_LOADING_FILES_WITH_OPENCV_FAILED                          = 'verifier.loading_files_with_opencv_failed'
+    VERIFIER_RUNNING_BLENDER_FAILED                                    = 'verifier.running_blender_failed'
+    VERIFIER_UNPACKING_ARCHIVE_FAILED                                  = 'verifier.unpacking_archive_failed'
```
Comment (kbeker) [inline concent_api/utils/constants.py]: Please delete spaces:)

Comment (rwrzesien) [reply concent_api/utils/constants.py]: If you really want I will remove them, but isn't it a good example when it increases readability ? @dybi what do you think ?

File: concent_api/verifier/tasks.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: No, currently it will overwrite.

File: concent_api/verifier/tasks.py (hunk 10 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: I don't see an option to set it in unpacking function, so I would have to check manually if the any file already exists. But just a line before I am doing `clean_directory`, so there should not be any files there. Do you think we need this extra check ?

Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: Added, it will now crash with raising exception, is that what you expected ? Or should I catch it first and log, or handle somehow ?

File: concent_api/verifier/tasks.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: I don't see an option to set it in unpacking function, so I would have to manually extract files list and check if any file already exists. But just few lines before I am doing `clean_directory`, so there should not be any files there. Do you think we need this extra check ?

Comment (cameel) [reply concent_api/verifier/tasks.py]: It's just a sanity check. It would be better to have this but we can skip it if it adds complexity or requires a lot of extra work in this particular case.

File: concent_api/verifier/tasks.py (hunk 10 above)
Comment (cameel) [reply concent_api/verifier/tasks.py]: The handler on the task is supposed to catch it, report to sentry and then reraise. You do not have to do anything else.

File: concent_api/verifier/tasks.py [hunk 14]
```diff
@@ -17,15 +20,20 @@
...
+from .utils import generate_upload_file_name
+from .utils import generate_verifier_storage_file_path
 from .utils import get_files_list_from_archive
 from .utils import prepare_storage_request_headers
 from .utils import run_blender
 from .utils import store_file_from_response_in_chunks
 from .utils import unpack_archive
 
 
+crash_logger = logging.getLogger('concent.crash')
```
[the remark this answers is not in the document]
Comment (cameel) [reply concent_api/verifier/tasks.py]: One is for reporting to sentry, the other one is for writing to the log.

File: concent_api/verifier/tasks.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cameel) [reply concent_api/verifier/tasks.py]: Like I said elsewhere, this case is for uploading the rendering result for audit purposes. If it fails, well, tough luck. We should not stop verification because of that.

File: concent_api/verifier/tasks.py (hunk 8 above)
Comment (cameel) [reply concent_api/verifier/tasks.py]: By "archive can't be extracted" I mean a damaged archive. I think `BadZipFile` covers that. Not stuff like permission errors, missing files, etc. Those things should crash the verifier.

**Comment by cameel** [comment]
I have added answers for @rwrzesien's questions about error handling. But honestly, it's pretty hard to give good, consistent answers when reviewing each case in isolation. I'm afraid that my answers might not be consistent. Please apply the requested changes or answer if it's still unclear or if my suggestions seem inconsistent.

But I think I'll need to look at the task as a whole at some point and review the error handling from that perspective if you want consistent answers :)

File: concent_api/verifier/tasks.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: Should it just raise exception or schedule `verification_result` task with error ?

Comment (cameel) [reply concent_api/verifier/tasks.py]: Both :)

Ideally you would raise an exception, that exception would be caught just before escaping the task and it would result in the error being reported to Sentry and `verification_result` being sent to the control cluster. One important thing is that if you schedule a `verification_result`, the task must not be retried any more.

Thats a bit like the decorator catching and reporting exceptions but not quite because it's specific to verification (it reports the result) and it does not reraise the exception.

Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: :+ done
